libsignal-service/src/provisioning/manager.rs

```diff
@@ -40,82 +46,33 @@ pub struct ConfirmCodeResponse {
     pub storage_capable: bool,
 }
 
-pub struct ProvisioningManager<'a, P: PushService + 'a> {
-    push_service: &'a mut P,
-    phone_number: PhoneNumber,
-    password: String,
-}
-
-#[derive(Derivative)]
-#[derivative(Debug)]
+#[derive(Debug)]
 pub enum SecondaryDeviceProvisioning {
     Url(Url),
-    NewDeviceRegistration {
-        phone_number: phonenumber::PhoneNumber,
-        device_id: DeviceId,
-        registration_id: u32,
-        pni_registration_id: u32,
-        service_ids: ServiceIds,
-        #[derivative(Debug = "ignore")]
-        aci_private_key: PrivateKey,
-        aci_public_key: PublicKey,
-        #[derivative(Debug = "ignore")]
-        pni_private_key: PrivateKey,
-        pni_public_key: PublicKey,
-        #[derivative(Debug = "ignore")]
-        profile_key: Vec<u8>,
-    },
+    NewDeviceRegistration(NewDeviceRegistration),
 }
 
-impl<'a, P: PushService + 'a> ProvisioningManager<'a, P> {
-    pub fn new(
-        push_service: &'a mut P,
-        phone_number: PhoneNumber,
-        password: String,
-    ) -> Self {
-        Self {
-            push_service,
-            phone_number,
-            password,
-        }
-    }
-
-    pub(crate) async fn confirm_device(
-        &mut self,
-        confirm_code: &str,
-        confirm_code_message: ConfirmDeviceMessage,
-    ) -> Result<DeviceId, ServiceError> {
-        self.push_service
-            .put_json(
-                Endpoint::Service,
-                &format!("/v1/devices/{}", confirm_code),
-                &[],
-                self.auth_override(),
-                confirm_code_message,
-            )
-            .await
-    }
-
-    fn auth_override(&self) -> HttpAuthOverride {
-        let credentials = ServiceCredentials {
-            uuid: None,
-            phonenumber: self.phone_number.clone(),
-            password: Some(self.password.clone()),
-            signaling_key: None,
-            device_id: None,
-        };
-        if let Some(auth) = credentials.authorization() {
-            HttpAuthOverride::Identified(auth)
-        } else {
-            HttpAuthOverride::NoOverride
-        }
-    }
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct NewDeviceRegistration {
+    pub phone_number: phonenumber::PhoneNumber,
+    pub device_id: DeviceId,
+    pub registration_id: u32,
+    pub pni_registration_id: u32,
+    pub service_ids: ServiceIds,
+    #[derivative(Debug = "ignore")]
+    pub aci_private_key: PrivateKey,
+    pub aci_public_key: PublicKey,
+    #[derivative(Debug = "ignore")]
+    pub pni_private_key: PrivateKey,
+    pub pni_public_key: PublicKey,
+    #[derivative(Debug = "ignore")]
+    pub profile_key: ProfileKey,
 }
 
 #[derive(Clone)]
 pub struct LinkingManager<P: PushService> {
```

Comment:
I think it makes sense to put the random generator also here inside.

Reply:
I actually would rather remove `LinkManager` since it's a struct that you never reuse, and I made the whole thing a free function called `link_device`.